src/bd/primitives/Specializer.java

```diff
@@ -0,0 +1,118 @@
+package bd.primitives;
+
+import java.lang.reflect.InvocationTargetException;
+
+import com.oracle.truffle.api.dsl.NodeFactory;
+import com.oracle.truffle.api.source.SourceSection;
+
+import bd.primitives.Primitive.NoChild;
+import bd.settings.VmSettings;
+
+
+/**
+ * A Specializer defines when a node can be used as a eager primitive, how
+ * it is to be instantiated, and acts as factory for them.
+ *
+ * @param <T> the type of the nodes created by this specializer
+ * @param <Context> the type of the context object
+ * @param <ExprT> the root type of expressions used by the language
+ */
+public class Specializer<T, Context, ExprT> {
+  protected final Context                    context;
+  protected final Primitive                  prim;
+  protected final NodeFactory<T>             fact;
+  private final NodeFactory<? extends ExprT> extraChildFactory;
+
+  @SuppressWarnings("unchecked")
+  public Specializer(final Primitive prim, final NodeFactory<T> fact, final Context context) {
+    this.prim = prim;
+    this.fact = fact;
+    this.context = context;
+
+    if (prim.extraChild() == NoChild.class) {
+      extraChildFactory = null;
+    } else {
+      try {
+        extraChildFactory =
+            (NodeFactory<? extends ExprT>) prim.extraChild().getMethod("getInstance")
+                                               .invoke(null);
+      } catch (IllegalAccessException | IllegalArgumentException
+          | InvocationTargetException | NoSuchMethodException
+          | SecurityException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  public boolean inParser() {
+    return prim.inParser() && !prim.requiresArguments();
+  }
+
+  public boolean noWrapper() {
+    return prim.noWrapper();
+  }
+
+  public boolean classSide() {
+    return prim.classSide();
+  }
+
+  public boolean matches(final Object[] args, final ExprT[] argNodes) {
+    // TODO: figure out whether we really want it like this with a VmSetting, or whether
+    // there should be something on the context
+    if (prim.disabled() && VmSettings.DYNAMIC_METRICS) {
+      return false;
+    }
+
```

Comment:
Ah, ok, I see why we need VMSettings now. But, I do not think is worth having at this moment...